File: choreolib/src/main/native/cpp/choreo/trajectory/ProjectFile.cpp
// Copyright (c) Choreo contributors

#include "choreo/trajectory/ProjectFile.h"

#include <string>

#include <wpi/json.h>

using namespace choreo;

void choreo::to_json(wpi::json& json, const Expression& exp) {
  json = wpi::json{{"exp", exp.expression}, {"val", exp.val}};
}

void choreo::from_json(const wpi::json& json, Expression& exp) {
  if (json.is_object() && json.size() == 2 && json["exp"].is_string() &&
      json["val"].is_number()) {
    exp.expression = json["exp"].get<std::string>();
    exp.val = json["val"].get<double>();
  } else {
    throw std::runtime_error("Invalid measurement format");
  }
}

void choreo::to_json(wpi::json& json, const XYExpression& xyexp) {
  json = wpi::json{{"x", xyexp.x}, {"y", xyexp.y}};
}
void choreo::from_json(const wpi::json& json, XYExpression& xyexp) {
  json.at("x").get_to(xyexp.x);
  json.at("y").get_to(xyexp.y);
}

void choreo::to_json(wpi::json& json, const Pose& pose) {
  json = wpi::json{{"x", pose.x}, {"y", pose.y}, {"heading", pose.heading}};
}

void choreo::from_json(const wpi::json& j, Pose& pose) {
  j.at("x").get_to(pose.x);
  j.at("y").get_to(pose.y);
  j.at("heading").get_to(pose.heading);
}

void choreo::to_json(wpi::json& json, const Bumpers& bumpers) {
  json = wpi::json{
      {"front", bumpers.front}, {"back", bumpers.back}, {"side", bumpers.side}};
}

void choreo::from_json(const wpi::json& json, Bumpers& bumpers) {
  json.at("front").get_to(bumpers.front);
  json.at("back").get_to(bumpers.back);
  json.at("side").get_to(bumpers.side);
}

void choreo::to_json(wpi::json& json, const Variable& var) {
  json = wpi::json{{"dimension", var.dimension}, {"var", var.var}};
}

void choreo::from_json(const wpi::json& json, Variable& var) {
  json.at("dimension").get_to(var.dimension);
  json.at("var").get_to(var.var);
}

void choreo::to_json(wpi::json& json, const Config& config) {
  json = wpi::json{{"frontLeft", config.frontLeft},
                   {"backLeft", config.backLeft},
                   {"mass", config.mass},
                   {"inertia", config.inertia},
                   {"gearing", config.gearing},
                   {"radius", config.wheelRadius},
                   {"vmax", config.vmax},
                   {"tmax", config.tmax},
                   {"bumper", config.bumpers},
                   {"differentialTrackWidth", config.differentialTrackWidth}};
}

void choreo::from_json(const wpi::json& json, Config& config) {
  json.at("frontLeft").get_to(config.frontLeft);
  json.at("backLeft").get_to(config.backLeft);
  json.at("mass").get_to<Expression>(config.mass);
  json.at("inertia").get_to(config.inertia);
  json.at("gearing").get_to(config.gearing);
  json.at("radius").get_to(config.wheelRadius);
  json.at("vmax").get_to(config.vmax);
  json.at("tmax").get_to(config.tmax);
  json.at("bumper").get_to(config.bumpers);
  json.at("differentialTrackWidth").get_to(config.differentialTrackWidth);
}

void choreo::to_json(wpi::json& json, const ProjectFile& projectFile) {
  json = wpi::json{
      {"name", projectFile.name},
      {"version", projectFile.version},
      {"type", projectFile.type},
      {"variables", wpi::json{{"expressions", projectFile.expressions},
                              {"poses", projectFile.poses}}},
      {"config", projectFile.config},
      {"generationFeatures", wpi::json::array()}};
}

void choreo::from_json(const wpi::json& json, ProjectFile& projectFile) {
  json.at("name").get_to(projectFile.name);
  json.at("version").get_to(projectFile.version);
  json.at("type").get_to(projectFile.type);

  const auto& variables = json.at("variables");
  if (variables.contains("expressions")) {
    for (const auto& [key, value] : variables["expressions"].items()) {
      projectFile.expressions[key] = value.get<Variable>();
    }
  }

  if (variables.contains("poses")) {
    for (const auto& [key, value] : variables["poses"].items()) {
      projectFile.poses[key] = value.get<Pose>();
    }
  }

  json.at("config").get_to(projectFile.config);
}

File: choreolib/src/main/java/choreo/trajectory/ProjectFile.java
// Copyright (c) Choreo contributors

package choreo.trajectory;

import java.util.List;

/** A representation of a project file aka a .chor. */
public class ProjectFile {
  /** A representation of an expression. An equation and its value. */
  public static class Expression {
    /** The equation. */
    public final String exp;

    /** The value. */
    public final double val;

    Expression(String exp, double val) {
      this.exp = exp;
      this.val = val;
    }
  }

  /** An xy pair of expressions. */
  public static class XYExpression {
    /** The x expression. */
    public final Expression x;

    /** The y expression. */
    public final Expression y;

    XYExpression(Expression x, Expression y) {
      this.x = x;
      this.y = y;
    }
  }

  /**
   * A collection of expressions representing the distance of the bumpers from the center of the
   * robot.
   */
  public static class Bumpers {
    /** The front bumper expression. */
    public final Expression front;

    /** The back bumper expression. */
    public final Expression back;

    /** The side bumper expression. */
    public final Expression side;

    Bumpers(Expression front, Expression back, Expression side) {
      this.front = front;
      this.back = back;
      this.side = side;
    }
  }

  /** The user configuration of the project. */
  public static class Config {
    /** The position of the front left module */
    public final XYExpression frontLeft;

    /** The position of the back left module */
    public final XYExpression backLeft;

    /** The mass of the robot. (kg) */
    public final Expression mass;

    /** The inertia of the robot. (kg-m^2) */
    public final Expression inertia;

    /** The gearing of the robot. */
    public final Expression gearing;

    /** The radius of the wheel. (m) */
    public final Expression wheelRadius;

    /** The maximum velocity of the robot. (m/s) */
    public final Expression vmax;

    /** The maximum torque of the robot. (N-m) */
    public final Expression tmax;

    /** The bumpers of the robot. */
    public final Bumpers bumper;

    /** The width between the wheels of the robot. (m) */
    public final Expression differentialTrackWidth;

    Config(
        XYExpression frontLeft,
        XYExpression backLeft,
        Expression mass,
        Expression inertia,
        Expression gearing,
        Expression wheelRadius,
        Expression vmax,
        Expression tmax,
        Bumpers bumper,
        Expression differentialTrackWidth) {
      this.frontLeft = frontLeft;
      this.backLeft = backLeft;
      this.mass = mass;
      this.inertia = inertia;
      this.gearing = gearing;
      this.wheelRadius = wheelRadius;
      this.vmax = vmax;
      this.tmax = tmax;
      this.bumper = bumper;
      this.differentialTrackWidth = differentialTrackWidth;
    }
  }

  /** The name of the project. */
  public final String name;

  /** The version of the project. */
  public final String version;

  /** The sample type for the project */
  public final String type;

  /** The configuration of the project. */
  public final Config config;

  /** The generation features of the project. */
  public final List<String> generationFeatures;

  ProjectFile(
      String name, String version, String type, Config config, List<String> generationFeatures) {
    this.name = name;
    this.version = version;
    this.type = type;
    this.config = config;
    this.generationFeatures = generationFeatures;
  }
}
